### PR TITLE
Handle CommonJS module

### LINKED ISSUE #868: sveltekit complains about using synthentic commonjs import
"svelte-vega": "^2.1.0",
"@sveltejs/kit": "^1.22.6",

```
import { vega } from "vega-embed";
         ^^^^
SyntaxError: Named export 'vega' not found. The requested module 'vega-embed' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:

import pkg from 'vega-embed';
const { vega } = pkg;
```

see https://github.com/vega/svelte-vega/blob/acc0b2504f4a7e7a40020f695c7d0dd7c5defc9b/packages/svelte-vega/src/lib/utils.ts#L2

### PR DESCRIPTION
**Pull request by liamdiprose**
Fixes #868

The `vega-embed` package is not of `"type": "module"`, so it has to be imported this way.

It seems to work for me in a stripped-down version of this package using Vite 5.

### CODE REVIEW

**Comment by domoritz** [comment]
Can you bump the vite dependency in this pull request?

**Comment by liamdiprose** [comment]
I'll give it a go on Monday. Enjoy your weekend!

**Comment by liamdiprose** [comment]
Hey, I checked the examples/storybook as you asked and found they didn't work. I needed to change it to this `import * as pkg` import pattern in order to get it working for both my project and this project's examples.

**Comment by liamdiprose** [comment]
I haven't added the `yarn.lock` file. Do you want me to include it?

**Comment by domoritz** [comment]
Yes. We should make the ci fail if the lock file is not up to date.

**Comment by liamdiprose** [comment]
Apparently `yarn --frozen-lockfile` is supposed to error if it isn't up to date, and yet "all checks have passed".

**Comment by domoritz** [comment]
Yeah, I expected that to error out. Weird.

**Comment by domoritz** [comment]
Hmm, does storybook work for you? I am seeing this error on this branch. 

Screenshot 2024-02-12 at 14 45 44

**Comment by liamdiprose** [comment]
My bad - thanks for catching this. Turns out `import * from ...` pulls in a bunch of stuff meant for node only. I've revisited the import line again, and now we have something before this PR, but the `*.module.js` file is resolved manually.

Storybook, Samples, and my project work now (for me).

**Comment by domoritz** [comment]
Hmm, why doesn't vite pull in the module file automatically? We already specify module in the package.json. Maybe we need to add an explicit export map?

**Comment by domoritz** [comment]
Hmm, looks like vite prefers browser over module: https://vitejs.dev/config/shared-options#resolve-mainfields. Not sure why, though.

**Comment by domoritz** [comment]
Adding explicit exports for embed seems to work (https://github.com/vega/vega-embed/pull/1301) but now vite doesn't like fast-json-patch which again to me doesn't make sense since the package is esm as well.

```
[vite] Named export 'applyPatch' not found. The requested module 'fast-json-patch' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:

import pkg from 'fast-json-patch';
const {applyPatch} = pkg;
```

File: packages/svelte-vega/src/lib/utils.ts [hunk 1]
```diff
@@ -1,5 +1,5 @@
 import type { View, SignalListeners } from './types';
-import { vega } from 'vega-embed';
+import { vega } from 'vega-embed/build/vega-embed.module.js';
```
Comment (domoritz) [inline packages/svelte-vega/src/lib/utils.ts:2]: That doesn't feel right as it requires a specific file in a specific place.

The right fix seem to add import maps.

**Comment by domoritz** [comment]
This branch isn't working for me btw. This is the svelte kit example.

Screenshot 2024-02-12 at 22 08 05

**Comment by liamdiprose** [comment]
Yeah, I'm getting the same error as you for sveltekit. Svelte works, which suggests that vite is resolving it correctly, but node isn't. 

Storybook was still broken, but I got it to work by setting this `resolve.alias`: https://github.com/vitejs/vite/discussions/9330#discussioncomment-5776373 . This is hacky, because its caused by a browser trying to import a nodejs-only module (node-fetch). It would be better to stop it happening in the first place.

I wonder if node expects `require()` for CJS dependencies (vega-embed). Maybe there is a vite plugin and/or polyfill for this?

I had a go at getting vite to resolve the path to `vega-embed.module.js` using `resolve.alias`, but had no luck. I understand why you don't want to assume the file structure of a dependency, but without changing it to `"type": "module"`, I'm pessimistic that node will find it properly. If we must resolve the module's path manually here, then wouldn't it be best to leave it in the most obvious place so it can be debugged easier? (opposed to hiding it as a build-system rewrite plugin)

In summary, its getting messy. I'm going to revisit this in a day.

**Comment by domoritz** [comment]
I'm happy to change embed to use `exports` to convince Vite that Vega embed is an esm dependency.

**Comment by liamdiprose** [comment]
Sorry the stalled progress. I've got a carbon-copy of the svelte-vega package working with the initial commit of this PR.

While I can appreciate the structure of this repo, I can't justify working on Storybook when we only need a svelte-friendly wrapper for vegalite.

**Comment by domoritz** [comment]
Makes sense. I think storybook isn't the issue big here but svelte kit doesn't work either.

I'll close the pull request for now but it's still something we should address since many people use vite 5.
